google/cloud/pubsub/internal/tracing_batch_sink_test.cc

```diff
@@ -0,0 +1,504 @@
+    std::vector<opentelemetry::nostd::shared_ptr<opentelemetry::trace::Span>>
+        spans,
+    std::shared_ptr<BatchSink> const& batch_sink, bool end_spans = true) {
+  for (size_t i = 0; i < spans.size(); i++) {
+    auto message =
+        pubsub::MessageBuilder().SetData("test" + std::to_string(i)).Build();
+    auto scope = opentelemetry::trace::Scope(spans[i]);
+    batch_sink->AddMessage(message);
+    if (end_spans) {
+      spans[i]->End();
+    }
+  }
+}
+
+/// Makes test options.
+auto MakeTestOptions(size_t max_otel_link_count = kDefaultMaxLinks) {
+  Options options;
+  options.set<pubsub::MaxOtelLinkCountOption>(max_otel_link_count);
+  return options;
+}
+
+std::shared_ptr<BatchSink> MakeTestBatchSink(std::shared_ptr<BatchSink> mock) {
+  return MakeTracingBatchSink(std::move(mock), MakeTestOptions());
+}
+
+pubsub::Topic TestTopic() {
+  return pubsub::Topic("test-project", "test-topic");
+}
+
+google::pubsub::v1::PublishRequest MakeRequest(int n) {
+  google::pubsub::v1::PublishRequest request;
+  request.set_topic(TestTopic().FullName());
+  for (int i = 0; i != n; ++i) {
+    request.add_messages()->set_message_id("message-" + std::to_string(i));
+  }
+  return request;
+}
+
+google::pubsub::v1::PublishResponse MakeResponse(
+    google::pubsub::v1::PublishRequest const& request) {
+  google::pubsub::v1::PublishResponse response;
+  for (auto const& m : request.messages()) {
+    response.add_message_ids("id-" + m.message_id());
+  }
+  return response;
+}
+
+TEST(TracingBatchSink, AddMessageAddsEvent) {
+  auto span_catcher = InstallSpanCatcher();
+  auto span = MakeSpan("test span");
+  opentelemetry::trace::Scope scope(span);
+  auto mock = std::make_shared<pubsub_testing::MockBatchSink>();
+  EXPECT_CALL(*mock, AddMessage(_));
+  auto batch_sink = MakeTestBatchSink(std::move(mock));
+
+  auto message = pubsub::MessageBuilder().SetData("test").Build();
+
+  batch_sink->AddMessage(message);
+
+  span->End();
+
+  EXPECT_THAT(
+      span_catcher->GetSpans(),
+      Contains(AllOf(SpanHasEvents(EventNamed("gl-cpp.added_to_batch")))));
+}
+
+TEST(TracingBatchSink, AsyncPublish) {
+  namespace sc = ::opentelemetry::trace::SemanticConventions;
+  auto span_catcher = InstallSpanCatcher();
+  auto message_span = MakeSpan("test span");
+  auto mock = std::make_unique<pubsub_testing::MockBatchSink>();
+  EXPECT_CALL(*mock, AddMessage(_));
+  EXPECT_CALL(*mock, AsyncPublish)
+      .WillOnce([](google::pubsub::v1::PublishRequest const& request) {
+        EXPECT_TRUE(ThereIsAnActiveSpan());
+        EXPECT_TRUE(OTelContextCaptured());
+        EXPECT_THAT(request, IsProtoEqual(MakeRequest(1)));
+        return make_ready_future(make_status_or(MakeResponse(request)));
+      });
+
+  auto batch_sink = MakeTestBatchSink(std::move(mock));
+
+  auto initial_spans = {message_span};
+  AddMessages(initial_spans, batch_sink);
+
+  auto response = batch_sink->AsyncPublish(MakeRequest(1)).get();
+  ASSERT_THAT(response, IsOk());
```

Comment:
nit: If we don't care about what is in `response`, we should prefer `EXPECT_THAT(...)` so execution will continue if the expectation fails.

Here and below.